### PR TITLE
bpo-30699: Improve example on tzinfo instances

### PR DESCRIPTION
**Pull request by mariocj89**
Move from GMTX to TZX when naming the classes, as GMT1 might be rather confusing as seen in the reported issue.
In addition, move to UTC over GMT and improve the tzname implementation.

https://bugs.python.org/issue30699

### CODE REVIEW

**Review by vstinner — APPROVED** [review]
LGTM.

**Comment by Mariatta** [comment]
Appveyor was stuck. I'll close and re-open the PR to retrigger it.

**Review by Mariatta — APPROVED** [review]

**Comment by Mariatta** [comment]
Thanks, the changes LGTM. If no objection from @abalkin I'll merge it.

**Comment by mariocj89** [comment]
Thanks a lot for reviewing this old PRs :) really appreciated

**Comment by abalkin** [comment]
If we are going to update these examples, I think we should make them PEP 495 compliant and I don't think this PR achieves it.

**Comment by mariocj89** [comment]
@abalkin not sure I understand what you mean. The examples are moved form UTC naming to GMT, there is no geographical timezone involved and therefore the fold attribute is not relevant. Isn't it?

**Comment by abalkin** [comment]
@mariocj89 - the fold attribute is only irrelevant for a TZ with a never changing offset.  As soon as you have DST changes as is the case in these examples, you have to take fold into account when implementing the `.utcoffset()` and `.dst()` methods.  See [Guidelines for New tzinfo Implementations](https://www.python.org/dev/peps/pep-0495/#id43).

**Comment by mariocj89** [comment]
@abalkin absolutely right, missed the `UTC +2" if self.dst(dt) else "UTC +1` in the exmple. I'll have a look into updating tbe PR to include that

**Review by abalkin — DISMISSED** [review]
Let's make the examples [PEP 495](https://www.python.org/dev/peps/pep-0495) compliant.

**Comment by pganssle** [comment]
@abalkin I trust Mario to come up with something here, but I do think writing a fold-aware example implementation would be way easier if [this issue](https://bugs.python.org/issue28602) were resolved. I think that would also make `fromutc` implementations that wrap `tzinfo.fromutc` much faster, since most of the code would be executed in C (though I could be wrong on that point).

I think the fact that it's actually fairly difficult to write a `fromutc` that's simple enough for a documentation example might be a sign that we should try to alleviate that cognitive overhead by having basic `fold` support in `tzinfo.fromutc`.

**Comment by mariocj89** [comment]
I am currently quite overwhelmed with work and other organisations, but as soon as I have time i’d Like to get into it and update the examples. I realised when doing it that it is not a 10 minutes job :p

TL;DR
I haven’t forgotten, will get into it “soon”

**Comment by mariocj89** [comment]
@abalkin The code for the timezone classes with support for dst and fold attributes is rather complex. 

Given the section is to illustrate how to work with datetime and tzinfo objects, what about moving to timezone.utc and a user-defined timezone class based on Kabul? I chose Kabul as it moved its offset once but has no ambiguous times (only imaginary).

This still showcases how datetime and tzinfo interact and the code is much simpler!

Note there are examples on how to work with tzinfo further down the file.

Example implementation: https://github.com/mariocj89/cpython/commit/c13263068049e2f296e0227e66c1501d5346e5b2

I can push this to this PR for review if the idea looks good.

**Comment by abalkin** [comment]
@mariocj89 - Using Asia/Kabul for the documentation example sounds fine, but your implementation is incomplete.  For the "imaginary" times the result of utcoffset() must depend on the fold value.  See [Mind the Gap section in PEP 495](https://www.python.org/dev/peps/pep-0495/#mind-the-gap).

**Comment by mariocj89** [comment]
Good point! Always thought that  the PEP referred only to DST but indeed apply to all “variable offsets”. 

Will then update it and push it to this branch. Thanks.

**Comment by pganssle** [comment]
As a note, if something like #7425 (which adds `fold` support directly into the default `fromutc` implementation) were to be merged, you could write a super-basic fold-aware implementation like this:

```python
from datetime import datetime, timedelta, tzinfo, timezone

UTC = timezone.utc

class ExampleFold(tzinfo):
    def tzname(self, dt):
        if self.dst(dt):
            return "DST"
        else:
            return "STD"

    def utcoffset(self, dt):
        return timedelta(hours=0) + self.dst(dt)

    def dst(self, dt):
        dt_start = datetime(1, 3, 17, 1).replace(year=dt.year)
        dt_end = datetime(1, 11, 4, 1).replace(year=dt.year)

        naive_dt = dt.replace(tzinfo=None)

        dst_offset = timedelta(hours=1)
        if dt_start <= naive_dt < dt_end:
            return dst_offset
        else:
            if naive_dt < dt_end + dt_offset:
                if dt.fold:
                    return timedelta(0)
                else:
                    return dst_offset

            return timedelta(0)
```

Obviously that's a sorta weird case because there's no timezone that uses fixed days of the year for offsets, but you can see how easy the "add fold support" part is once `fromutc` respects it.

**Comment by mariocj89** [comment]
@abalkin  I've pushed a potential implementation. Had to override `fromutc` as well, maybe there is a simpler way :/ Let me know if anything needs changing.

Also updated the content to pass doctests (only the modified part).

Tested that Kabul implementation against the dateutil one before at and after the move.

File: Doc/library/datetime.rst [hunk 1]
```diff
@@ -1316,56 +1316,58 @@ Examples of working with datetime objects:
...
     ...         else:
-    ...             return timedelta(0)
+    ...             return dt + timedelta(hours=4)
+    ...
+    ...     def dst(self, dt):
+    ...         return timedelta(0)
+    ...
     ...     def tzname(self,dt):
-    ...         return "GMT +2"
+    ...         return "Asia/Kabul"
```
Comment (abalkin) [inline Doc/library/datetime.rst]: I think this should return UTC+04:30 / UTC+04 depending on the time relative to the move date.

Comment (mariocj89) [reply Doc/library/datetime.rst]: From the docs:

> Return the time zone name corresponding to the datetime object dt, as a string. Nothing about string names is defined by the datetime module, and there’s no requirement that it mean anything in particular. For example, “GMT”, “UTC”, “-500”, “-5:00”, “EDT”, “US/Eastern”, “America/New York” are all valid replies

If the method is to return the timezone *name* I think it is probably better to have Kabul rather than the string representation of the offset. That said, both dateutil and pytz are returning the string representation of the offset, not sure why.

Happy to change it if you still think it is more correct.

Maybe @pganssle can reason why it is the offset rather than the name in dateutil.

Comment (abalkin) [reply Doc/library/datetime.rst]: Traditionally, tzname() is returning something similar to strftime('%Z').  In the case of Kabul, this means

```
$ TZ=Asia/Kabul date -r 1000000000 +%Z
+0430
```
for recent times, and
```
$ TZ=Asia/Kabul date -r -1000000000 +%Z
+04
```
for times before the transition.  Paul Eggert, has been leading a [crusade on "invented" timezone abbreviations](https://mm.icann.org/pipermail/tz/2017-February/024837.html) recently resulting in %Z format becoming useless for many parts of the World.  I still believe tzname() should return a time-dependent abbreviation rather than a permanent zone name.

Comment (mariocj89) [reply Doc/library/datetime.rst]: OK! Updated via https://github.com/python/cpython/pull/4290/commits/56e7adfe703d8eb0b165eeeb131d27d86c9ccbee

**Comment by vstinner** [comment]
I removed the " needs backport to 3.6" label, the 3.6 branch no longer accept bugfixes (only security fixes are accepted): https://devguide.python.org/#status-of-python-branches

File: Doc/library/datetime.rst [hunk 2]
```diff
@@ -1316,56 +1316,61 @@ Examples of working with datetime objects:
...
+    >>> dt1 = datetime(1900, 11, 21, 16, 30, tzinfo=tz1)
     >>> dt1.utcoffset()
-    datetime.timedelta(seconds=3600)
-    >>> dt2 = datetime(2006, 6, 14, 13, 0, tzinfo=gmt1)
-    >>> dt2.dst()
-    datetime.timedelta(seconds=3600)
+    datetime.timedelta(0, 14400)
+    >>> # Datetime after the change
+    >>> dt2 = datetime(2006, 6, 14, 13, 0, tzinfo=tz1)
     >>> dt2.utcoffset()
```
Comment (pganssle) [inline Doc/library/datetime.rst]: Can we change this to:

```python
>>> print(dt.utcoffset())
4:30:00
```

It's a very minor thing, but to me the standard `repr` for `timedelta` is completely unreadable.

File: Doc/library/datetime.rst [hunk 3]
```diff
@@ -1316,56 +1316,61 @@ Examples of working with datetime objects:
...
+    ...         return timedelta(0)
+    ...
+    ...     def tzname(self, dt):
+    ...         if dt >= self.UTC_MOVE_DATE:
+    ...             return "+04:30"
     ...         else:
-    ...             return timedelta(0)
-    ...     def tzname(self,dt):
-    ...         return "GMT +2"
+    ...             return "+04"
```
Comment (pganssle) [inline Doc/library/datetime.rst:1387]: I would mildly prefer `"+04:00"` here for consistency, but I don't care strongly about this, it's just a documentation example and I do realize that @abalkin suggested this label.

File: Doc/library/datetime.rst [hunk 4]
```diff
@@ -1316,56 +1316,61 @@ Examples of working with datetime objects:
 
 Using datetime with tzinfo:
 
-    >>> from datetime import timedelta, datetime, tzinfo
-    >>> class GMT1(tzinfo):
+    >>> from datetime import timedelta, datetime, tzinfo, timezone
+    >>> class KabulTz(tzinfo):
```
Comment (pganssle) [inline Doc/library/datetime.rst:1350]: If you add:

```python
def __repr__(self):
    return f"{self.__class__.__name__}()"
```
The later examples where you attach this to a datetime will be easier to read, e.g.:

```python
   >>> dt2     # doctest: +ELLIPSIS
   datetime.datetime(2006, 6, 14, 13, 0, tzinfo=KabulTz())
```

You may also consider adding an `__str__` that returns the tzdata IANA key (though I believe this is used nowhere in your examples later):

```python
def __str__(self):
    return "Asia/Kabul"
```

File: Doc/library/datetime.rst [hunk 5]
```diff
@@ -1591,26 +1596,25 @@ Instance methods:
...
     >>> from datetime import time, tzinfo, timedelta
-    >>> class GMT1(tzinfo):
+    >>> class TZ1(tzinfo):
     ...     def utcoffset(self, dt):
     ...         return timedelta(hours=1)
     ...     def dst(self, dt):
     ...         return timedelta(0)
     ...     def tzname(self,dt):
-    ...         return "Europe/Prague"
+    ...         return "UTC+1"
```
Comment (pganssle) [inline Doc/library/datetime.rst]: Super annoyingly, [POSIX-style offset names](https://github.com/dateutil/dateutil/issues/70#issuecomment-88982122) have an inverted sense, so `dateutil.parser.parse` (and I think `TZ` variables) would interpret `UTC+1` as the same as `-01:00`. Of course, many people use this in the opposite sense, which gives rise to some confusion.

I recommend bypassing the whole thing by returning "+01:00" here.

File: Doc/library/datetime.rst [hunk 6]
```diff
@@ -1591,26 +1596,25 @@ Instance methods:
 Example:
 
     >>> from datetime import time, tzinfo, timedelta
-    >>> class GMT1(tzinfo):
+    >>> class TZ1(tzinfo):
```
Comment (pganssle) [inline Doc/library/datetime.rst:1648]: As with above, a `__repr__` would be a good idea here.

**Review by pganssle — CHANGES_REQUESTED** [review]
Almost all of my comments here are completely optional and I think this can be merged whether they are incorporated or not.

The only "hard" change that I recommend is the use of `"+01:00"` instead of `"UTC+1"` in the `TZ1` class.

**Comment by csabella** [comment]
@mariocj89 @pganssle 
There was a lot of activity on this issue and 2 approvals, but I'm not quite sure where it's at.  Can you take another look and comment if it's ready to merge?  Thanks!

**Comment by mariocj89** [comment]
Wow! Totally forgot about this PR 😆 .

@abalkin [requested to make this PEP 495 complient](https://github.com/python/cpython/pull/4290#pullrequestreview-100919241).

There were some suggestions by @pganssle as well but were the opposite of what @abalkin was proposing when formatting offsets.

Let me know if there is any change to make!

**Comment by pganssle** [comment]
We can ignore any of my comments that contradict @abalkin's. This is clearly an improvement over the previous state of affairs, and we can always change it later.

**Comment by vstinner** [comment]
@pganssle: You didn't approve this PR. Does it mean that it still lacks someone, or is it good as it is? If yes, can you please officially approve it?

**Review by pganssle — APPROVED** [review]
@vstinner No problem, approved. I think this change is clearly an improvement over the status quo.

I think documentation backports are only going back to 3.7 now anyway, but in case they aren't, this relies on PEP 495, so it can't be backported any further than 3.6. I think it's fine to backport to 3.7 unless @mariocj89 disagrees.

**Comment by vstinner** [comment]
> I think documentation backports are only going back to 3.7 now anyway, but in case they aren't, this relies on PEP 495, so it can't be backported any further than 3.6. I think it's fine to backport to 3.7 unless @mariocj89 disagrees.

Python 3.6 no longer accept doc changes, only security fixes: https://devguide.python.org/#status-of-python-branches

**Comment by vstinner** [comment]
Thanks Mario Corchero for the doc enhancement, and Paul for the review!
